contribute-to-tidb/review-a-pr-stage.md

```diff
@@ -0,0 +1,238 @@
+
+Basically, PR authors are supposed to sufficiently test their PRs before code reviews. However, as a reviewer, you should still think about edge cases, especially for problems about concurrency.
+
+You can always validate the PR if you want. 
+
+For some changes such as UI, it is hard to understand how the changes will impact users when you are just reading the code. In this case, if it is too inconvenient to try it yourself, you can ask the PR author to give you a demo of the functionality.
+
+### Logic
+
+When concurrent programming is involved, it is particularly important to review the logic carefully, because the concurrent issues are very hard to detect by just running the code.
+
+### Complexity
+
+Is the PR more complex than it should be? To learn that, you can check the following items when reviewing the complexity of the PR:
+
+* Are individual lines too complex?
+* Are functions too complex?
+* Are classes too complex?
+
+"Too complex" usually means "the code changes cannot be understood quickly by readers" or "PR authors are likely to introduce bugs when they try to call or modify the code".
+
+A particular type of complexity is over-engineering, which means that PR authors have made the code more generic than it needs to be, or added functionality that is not presently needed by the system. Reviewers should be especially vigilant about over-engineering. The TiDB community encourages PR authors to solve the problems that must be solved now -- not the problems that might need to be solved in the future. 
+
+### Tests
+
+A PR should be test covered, whether the tests are unit tests, integration tests, or end-to-end tests.
+
+To ensure the tests in the PR are correct, sensible, and useful, you can check the following:
+
+* Whether the tests cover all interesting cases?
+* Are the test cases readable?
+* Does the PR lower the overall test coverage? 
+
+To ensure the tests in the PR are sufficient, you can simulate cases where the code may break and check the following:
+
+* Will the tests fail when the code is broken?
+* Does each test make simple and useful assertions?
+* Are the tests separated appropriately between different test methods?
+
+Remember that tests are also code that have to be maintained, and thus do not accept unnecessary complexity in tests.
+
+Last but not least, keep an eye on how long a single unit takes to run. That is, determine whether the test slow down the CI.
+
+### Naming
+
+Does the developer pick good names for everything? A good name is long enough to fully communicate what the item is or does, without being so long that it becomes hard to read.
+
+### Comments
+
+* Does the developer write clear comments in understandable English?
+* Are all of the comments actually necessary?
+
+Usually, comments are useful when they explain why the code exists, while not explain what the code is doing. If the code isn't clear enough to explain itself, then the code should be simplified.
```

Comment:
```suggestion
Usually, comments are useful to explain why the code exists, while not useful to explain what the code is doing. If the code is not clear enough to explain itself, the PR author should simplify the code instead of adding unnecessary comments.
```